File: batch_base64.py


### 1. `batch_base64.py` (Python)

```python
# batch_base64.py — Python версия

import os
import sys
import base64
import json
import argparse
import time
from pathlib import Path
from concurrent.futures import ThreadPoolExecutor, as_completed
from datetime import datetime
from colorama import init, Fore, Style

init(autoreset=True)

class BatchBase64:
    def __init__(self, input_dir, output_dir=None, threads=4, decode=False, recursive=True, report=True):
        self.input_dir = Path(input_dir)
        self.output_dir = Path(output_dir) if output_dir else self.input_dir / "base64_output"
        self.threads = threads
        self.decode = decode
        self.recursive = recursive
        self.report = report
        self.stats = {"processed": 0, "errors": 0, "failed": [], "start_time": None, "end_time": None}
        self.output_dir.mkdir(parents=True, exist_ok=True)

    def get_files(self):
        """Собирает все файлы для обработки."""
        pattern = "**/*" if self.recursive else "*"
        files = list(self.input_dir.glob(pattern))
        return [f for f in files if f.is_file()]

    def process_file(self, file_path):
        """Обрабатывает один файл (кодирует или декодирует)."""
        try:
            rel_path = file_path.relative_to(self.input_dir)
            if self.decode:
                # Декодирование: читаем .b64 файл, сохраняем исходник
                output_name = rel_path.with_suffix('')  # убираем .b64
                output_path = self.output_dir / output_name
                output_path.parent.mkdir(parents=True, exist_ok=True)
                with open(file_path, 'r') as f:
                    encoded = f.read()
                decoded = base64.b64decode(encoded)
                with open(output_path, 'wb') as f:
                    f.write(decoded)
            else:
                # Кодирование
                output_name = rel_path.with_suffix(rel_path.suffix + '.b64')
                output_path = self.output_dir / output_name
                output_path.parent.mkdir(parents=True, exist_ok=True)
                with open(file_path, 'rb') as f:
                    data = f.read()
                encoded = base64.b64encode(data).decode('ascii')
                with open(output_path, 'w') as f:
                    f.write(encoded)
            return True, str(rel_path), None
        except Exception as e:
            return False, str(rel_path), str(e)

    def run(self):
        """Запускает массовую обработку."""
        files = self.get_files()
        if not files:
            print(Fore.YELLOW + "⚠️ Нет файлов для обработки.")
            return

        print(Fore.CYAN + f"🔐 Base64 Batch Encoder (Python)")
        print(f"📁 Папка: {self.input_dir}")
        print(f"📂 Найдено {len(files)} файлов")
        print(f"⚡ Параллельная обработка ({self.threads} потоков)...")
        print()

        self.stats["start_time"] = datetime.now()
        processed = 0
        total = len(files)

        with ThreadPoolExecutor(max_workers=self.threads) as executor:
            futures = {executor.submit(self.process_file, f): f for f in files}
            for future in as_completed(futures):
                success, path, error = future.result()
                processed += 1
                if success:
                    self.stats["processed"] += 1
                else:
                    self.stats["errors"] += 1
                    self.stats["failed"].append({"path": path, "error": error})

                # Прогресс-бар
                percent = (processed / total) * 100
                bar_len = 30
                filled = int(bar_len * processed / total)
                bar = '█' * filled + '░' * (bar_len - filled)
                sys.stderr.write(f"\r[{bar}] {percent:.1f}% {processed}/{total}")
                sys.stderr.flush()

        self.stats["end_time"] = datetime.now()
        elapsed = (self.stats["end_time"] - self.stats["start_time"]).total_seconds()

        print()
        print()
        print(Fore.GREEN + f"✅ Обработано: {self.stats['processed']} файлов")
        print(Fore.YELLOW + f"⚠️ Ошибок: {self.stats['errors']}")
        print(Fore.CYAN + f"⏱️ Время: {elapsed:.2f} сек")
        print(Fore.GREEN + f"💾 Сохранено: {self.output_dir}")

        if self.report:
            self.save_report()

    def save_report(self):
        """Сохраняет отчёт в JSON."""
        report = {
            "input_dir": str(self.input_dir),
            "output_dir": str(self.output_dir),
            "total_files": self.stats["processed"] + self.stats["errors"],
            "processed": self.stats["processed"],
            "errors": self.stats["errors"],
            "failed": self.stats["failed"],
            "start_time": self.stats["start_time"].isoformat(),
            "end_time": self.stats["end_time"].isoformat(),
            "elapsed_seconds": (self.stats["end_time"] - self.stats["start_time"]).total_seconds(),
            "decode_mode": self.decode
        }
        report_path = self.output_dir / "report.json"
        with open(report_path, 'w', encoding='utf-8') as f:
            json.dump(report, f, indent=2, ensure_ascii=False)
        print(Fore.GREEN + f"📊 Отчёт: {report_path}")

def main():
    parser = argparse.ArgumentParser(description="Base64 Batch Encoder")
    parser.add_argument("input_dir", help="Папка с файлами для обработки")
    parser.add_argument("--output", "-o", help="Папка для сохранения результатов")
    parser.add_argument("--threads", "-t", type=int, default=4, help="Количество потоков")
    parser.add_argument("--decode", "-d", action="store_true", help="Декодировать файлы")
    parser.add_argument("--no-recursive", action="store_true", help="Не обрабатывать вложенные папки")
    parser.add_argument("--no-report", action="store_true", help="Не создавать отчёт")
    args = parser.parse_args()

    if not os.path.exists(args.input_dir):
        print(Fore.RED + f"❌ Папка не найдена: {args.input_dir}")
        sys.exit(1)

    batch = BatchBase64(
        args.input_dir,
        args.output,
        args.threads,
        args.decode,
        not args.no_recursive,
        not args.no_report
    )
    batch.run()

if __name__ == "__main__":
    main()

File: batch_base64.java
// batch_base64.java — Java версия

import java.io.*;
import java.nio.file.*;
import java.util.*;
import java.util.concurrent.*;
import java.util.Base64;
import java.time.*;

public class batch_base64 {
    private String inputDir;
    private String outputDir;
    private int threads;
    private boolean decode;
    private boolean recursive;
    private boolean report;
    private int processed = 0;
    private int errors = 0;
    private List<String[]> failed = new ArrayList<>();
    private long startTime, endTime;

    public batch_base64(String inputDir, String outputDir, int threads, boolean decode, boolean recursive, boolean report) {
        this.inputDir = inputDir;
        this.outputDir = outputDir != null ? outputDir : Paths.get(inputDir, "base64_output").toString();
        this.threads = threads;
        this.decode = decode;
        this.recursive = recursive;
        this.report = report;
    }

    private List<Path> getFiles() throws IOException {
        List<Path> files = new ArrayList<>();
        Path inputPath = Paths.get(inputDir);
        if (recursive) {
            Files.walk(inputPath).filter(Files::isRegularFile).forEach(files::add);
        } else {
            Files.list(inputPath).filter(Files::isRegularFile).forEach(files::add);
        }
        return files;
    }

    private void processFile(Path filePath) throws IOException {
        Path relPath = inputPath.relativize(filePath);
        String relPathStr = relPath.toString();

        Path outputPath;
        if (decode) {
            String name = relPathStr;
            if (name.endsWith(".b64")) {
                name = name.substring(0, name.length() - 4);
            }
            outputPath = Paths.get(outputDir, name);
        } else {
            outputPath = Paths.get(outputDir, relPathStr + ".b64");
        }
        Files.createDirectories(outputPath.getParent());

        if (decode) {
            String content = new String(Files.readAllBytes(filePath), "UTF-8");
            byte[] decoded = Base64.getDecoder().decode(content.trim());
            Files.write(outputPath, decoded);
        } else {
            byte[] data = Files.readAllBytes(filePath);
            String encoded = Base64.getEncoder().encodeToString(data);
            Files.write(outputPath, encoded.getBytes("UTF-8"));
        }
        processed++;
    }

    public void run() throws Exception {
        System.out.println("\u001B[36m🔐 Base64 Batch Encoder (Java)\u001B[0m");
        System.out.println("📁 Папка: " + inputDir);

        List<Path> files = getFiles();
        System.out.println("📂 Найдено " + files.size() + " файлов");
        System.out.println("⚡ Параллельная обработка (" + threads + " потоков)...\n");

        startTime = System.currentTimeMillis();
        ExecutorService executor = Executors.newFixedThreadPool(threads);
        List<Future<?>> futures = new ArrayList<>();

        for (Path file : files) {
            futures.add(executor.submit(() -> {
                try {
                    processFile(file);
                } catch (Exception e) {
                    errors++;
                    failed.add(new String[]{file.toString(), e.getMessage()});
                }
            }));
        }

        for (Future<?> future : futures) {
            try {
                future.get();
            } catch (Exception e) {
                // already handled
            }
        }
        executor.shutdown();
        executor.awaitTermination(60, TimeUnit.SECONDS);
        endTime = System.currentTimeMillis();
        double elapsed = (endTime - startTime) / 1000.0;

        System.out.println();
        System.out.println("\u001B[32m✅ Обработано: " + processed + " файлов\u001B[0m");
        System.out.println("\u001B[33m⚠️ Ошибок: " + errors + "\u001B[0m");
        System.out.println("\u001B[36m⏱️ Время: " + String.format("%.2f", elapsed) + " сек\u001B[0m");
        System.out.println("\u001B[32m💾 Сохранено: " + outputDir + "\u001B[0m");

        if (report) {
            saveReport();
        }
    }

    private void saveReport() throws IOException {
        Map<String, Object> report = new LinkedHashMap<>();
        report.put("input_dir", inputDir);
        report.put("output_dir", outputDir);
        report.put("total_files", processed + errors);
        report.put("processed", processed);
        report.put("errors", errors);
        report.put("failed", failed);
        report.put("start_time", Instant.ofEpochMilli(startTime).toString());
        report.put("end_time", Instant.ofEpochMilli(endTime).toString());
        report.put("elapsed_seconds", (endTime - startTime) / 1000.0);
        report.put("decode_mode", decode);

        String json = new com.google.gson.GsonBuilder().setPrettyPrinting().create().toJson(report);
        Files.write(Paths.get(outputDir, "report.json"), json.getBytes());
        System.out.println("\u001B[32m📊 Отчёт: " + outputDir + "/report.json\u001B[0m");
    }

    public static void main(String[] args) throws Exception {
        String inputDir = null;
        String outputDir = null;
        int threads = Runtime.getRuntime().availableProcessors();
        boolean decode = false;
        boolean recursive = true;
        boolean report = true;

        for (int i = 0; i < args.length; i++) {
            if (args[i].equals("--output") || args[i].equals("-o")) {
                outputDir = args[++i];
            } else if (args[i].equals("--threads") || args[i].equals("-t")) {
                threads = Integer.parseInt(args[++i]);
            } else if (args[i].equals("--decode") || args[i].equals("-d")) {
                decode = true;
            } else if (args[i].equals("--no-recursive")) {
                recursive = false;
            } else if (args[i].equals("--no-report")) {
                report = false;
            } else if (inputDir == null && !args[i].startsWith("-")) {
                inputDir = args[i];
            }
        }

        if (inputDir == null) {
            System.out.println("Usage: java batch_base64 <input_dir> [--output dir] [--threads N] [--decode]");
            System.exit(1);
        }

        if (!Files.exists(Paths.get(inputDir))) {
            System.out.println("\u001B[31m❌ Папка не найдена: " + inputDir + "\u001B[0m");
            System.exit(1);
        }

        batch_base64 batch = new batch_base64(inputDir, outputDir, threads, decode, recursive, report);
        batch.run();
    }
}
